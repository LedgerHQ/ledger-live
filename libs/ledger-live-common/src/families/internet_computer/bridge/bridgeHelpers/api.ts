import { log } from "@ledgerhq/logs";
import { AxiosRequestConfig, AxiosResponse } from "axios";
import { getEnv } from "@ledgerhq/live-env";
import network from "@ledgerhq/live-network/network";
import { ICP_BLK_NAME_ROSETTA, ICP_NET_ID_ROSETTA } from "../../consts";
import {
  ICPRosettaBlockHeightResponse,
  ICPRosettaGetBalancesResponse,
  ICPRosettaGetTxnsHistoryResponse,
} from "./icpRosetta/types";

const getICPURL = (path?: string): string => {
  const baseUrl = getEnv("API_ICP_ENDPOINT");
  if (!baseUrl) throw new Error("API base URL not available");

  return `${baseUrl}${path ? path : ""}`;
};

const ICPFetchWrapper = async <T>(path: string, body: any) => {
  const url = getICPURL(path);

  // We force data to this way as network func is not using the correct param type. Changing that func will generate errors in other implementations
  const opts: AxiosRequestConfig = {
    method: "POST",
    data: body,
    url,
  };
  const rawResponse = await network(opts);
  if (rawResponse && rawResponse.data && rawResponse.data.details?.error_message) {
    log("error", rawResponse.data.details?.error_message);
  }

  // We force data to this way as network func is not using the correct param type. Changing that func will generate errors in other implementations
  const { data } = rawResponse as AxiosResponse<T>;

  log("http", url);
  return data;
};

export const getICPRosettaNetworkIdentifier = () => {
  return {
    network_identifier: {
      blockchain: ICP_BLK_NAME_ROSETTA,
      network: ICP_NET_ID_ROSETTA,
    },
  };
};

export const fetchBlockHeight = async (): Promise<ICPRosettaBlockHeightResponse> => {
  const data = await ICPFetchWrapper<ICPRosettaBlockHeightResponse>(
    "network/status",
    getICPRosettaNetworkIdentifier(),
  );
  return data;
};

export const fetchBalances = async (accountId: string): Promise<ICPRosettaGetBalancesResponse> => {
  const body = {
    ...getICPRosettaNetworkIdentifier(),
    account_identifier: {
      address: accountId,
      metadata: {},
    },
  };
  const data = await ICPFetchWrapper<ICPRosettaGetBalancesResponse>("account/balance", body);
  return data;
};

export const fetchTxns = async (accountId: string): Promise<ICPRosettaGetTxnsHistoryResponse> => {
  const body = {
    ...getICPRosettaNetworkIdentifier(),
    account_identifier: {
      address: accountId,
      metadata: {},
    },
  };
  const data = await ICPFetchWrapper<ICPRosettaGetTxnsHistoryResponse>("search/transactions", body);
  return data;
};

export const constructionInvoke = async <TRequest, TResponse>(
  opts: TRequest,
  method: string,
): Promise<TResponse> => {
  const body: TRequest = {
    ...opts,
  };

  const data = await ICPFetchWrapper<TResponse>(`construction/${method}`, body);

  return data;
};
