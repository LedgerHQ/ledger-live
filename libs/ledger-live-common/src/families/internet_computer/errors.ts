import { createCustomErrorClass } from "@ledgerhq/errors";

/*
 * When the transferID/Memo is non number
 */
export const InvalidMemoICP = createCustomErrorClass("InvalidMemoICP");
