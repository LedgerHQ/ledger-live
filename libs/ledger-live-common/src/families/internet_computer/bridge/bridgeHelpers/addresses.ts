import { log } from "@ledgerhq/logs";
import { Account, Address } from "@ledgerhq/types-live";
import BigNumber from "bignumber.js";
import { MAX_MEMO_VALUE } from "../../consts";
import { fetchBalances } from "./api";

export const getAddress = (a: Account): Address =>
  a.freshAddresses.length > 0
    ? a.freshAddresses[0]
    : { address: a.freshAddress, derivationPath: a.freshAddressPath };

export async function validateAddress(address: string): Promise<{ isValid: boolean }> {
  try {
    const res = await fetchBalances(address);
    if (!res.balances) throw Error(res.details?.error_message);
    return { isValid: true };
  } catch (e: any) {
    log("error", e.message ?? "Failed to validate address");
    return { isValid: false };
  }
}

export function validateMemo(memo?: string): { isValid: boolean } {
  const res = BigNumber(memo ?? 0);

  if (res.isNaN() || res.lt(0) || res.gt(BigNumber(MAX_MEMO_VALUE))) {
    return { isValid: false };
  }

  return { isValid: true };
}
