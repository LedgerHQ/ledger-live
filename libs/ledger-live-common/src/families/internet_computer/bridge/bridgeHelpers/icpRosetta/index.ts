import Transport from "@ledgerhq/hw-transport";
import { Account } from "@ledgerhq/types-live";
import { Transaction } from "../../../types";
import ICP from "@zondax/ledger-icp";
import { constructionInvoke, getICPRosettaNetworkIdentifier } from "../api";
import {
  ICPRosettaConstructionCombineRequest,
  ICPRosettaConstructionCombineResponse,
  ICPRosettaConstructionHashRequest,
  ICPRosettaConstructionHashResponse,
  ICPRosettaConstructionPayloadsRequest,
  ICPRosettaConstructionPayloadsResponse,
  ICPRosettaConstructionSubmitRequest,
  ICPRosettaConstructionSubmitResponse,
  ICPRosettaConstructionDeriveRequest,
  ICPRosettaConstructionDeriveResponse,
} from "./types";
import { ingressExpiry, generateOperations, generateSignaturesPayload } from "./utils";
import { Cbor } from "@dfinity/agent";
import { Principal } from "@dfinity/principal";
import { isError } from "../../../utils";
import BigNumber from "bignumber.js";

export const getUnsignedTransaction = async (
  transaction: Transaction,
  account: Account,
): Promise<{
  unsignedTxn: string;
  payloads: ICPRosettaConstructionPayloadsResponse["payloads"];
}> => {
  const ops = generateOperations(transaction, account);
  const pubkeys = [
    {
      hex_bytes: account.xpub ?? "",
      curve_type: "secp256k1",
    },
  ];

  const reqOpts: ICPRosettaConstructionPayloadsRequest = {
    ...getICPRosettaNetworkIdentifier(),
    operations: ops,
    public_keys: pubkeys,
    metadata: {
      memo: parseInt(transaction.memo ?? "0"),
    },
  };
  const { payloads, unsigned_transaction } = await constructionInvoke<
    ICPRosettaConstructionPayloadsRequest,
    ICPRosettaConstructionPayloadsResponse
  >(reqOpts, "payloads");

  return { unsignedTxn: unsigned_transaction, payloads };
};

export const signICPTransaction = async ({
  unsignedTxn,
  transport,
  path,
  payloads,
  pubkey,
}: {
  unsignedTxn: string;
  transport: Transport;
  path: string;
  payloads: ICPRosettaConstructionPayloadsResponse["payloads"];
  pubkey: string;
}): Promise<{
  signatures: { txnSig: string; readSig: string };
  signedTxn: string;
}> => {
  const icp = new ICP(transport as any);
  const decodedTxn: any = Cbor.decode(Buffer.from(unsignedTxn, "hex"));
  const txnReqFromCbor = decodedTxn.updates[0][1];
  const expiry = new ingressExpiry(BigNumber(decodedTxn.ingress_expiries[0].toString()));

  const submitReq = {
    request_type: "call",
    canister_id: Principal.fromUint8Array(txnReqFromCbor.canister_id),
    method_name: txnReqFromCbor.method_name,
    arg: txnReqFromCbor.arg,
    sender: Principal.fromUint8Array(txnReqFromCbor.sender),
    ingress_expiry: expiry,
  };

  const txnBlobToSign = Cbor.encode({
    content: submitReq,
  });

  const signedTxnRes = await icp.sign(path, Buffer.from(txnBlobToSign), 0);
  isError(signedTxnRes);

  const result = {
    signatures: {
      readSig: "",
      txnSig: Buffer.from(signedTxnRes.signatureRS ?? "").toString("hex"),
    },
  };

  const signaturesPayload = generateSignaturesPayload(result.signatures, payloads, pubkey);

  const { signed_transaction: signedTxn } = await constructionInvoke<
    ICPRosettaConstructionCombineRequest,
    ICPRosettaConstructionCombineResponse
  >(
    {
      ...getICPRosettaNetworkIdentifier(),
      signatures: signaturesPayload,
      unsigned_transaction: unsignedTxn,
    },
    "combine",
  );

  return { ...result, signedTxn };
};

export const getTxnMetadata = async (signedTxn: string): Promise<{ hash: string }> => {
  const {
    transaction_identifier: { hash },
  } = await constructionInvoke<
    ICPRosettaConstructionHashRequest,
    ICPRosettaConstructionHashResponse
  >({ ...getICPRosettaNetworkIdentifier(), signed_transaction: signedTxn }, "hash");

  return { hash };
};

export const getTxnExpirationDate = (_unsignedTxn: string): Date => {
  return new Date();
};

export const broadcastTxn = async (signedTxn: string) => {
  await constructionInvoke<
    ICPRosettaConstructionSubmitRequest,
    ICPRosettaConstructionSubmitResponse
  >({ ...getICPRosettaNetworkIdentifier(), signed_transaction: signedTxn }, "submit");
};

export const deriveAddressFromPubkey = async (pubkey: string) => {
  const res = await constructionInvoke<
    ICPRosettaConstructionDeriveRequest,
    ICPRosettaConstructionDeriveResponse
  >(
    {
      ...getICPRosettaNetworkIdentifier(),
      public_key: { curve_type: "secp256k1", hex_bytes: pubkey },
    },
    "derive",
  );

  return res.account_identifier.address;
};
