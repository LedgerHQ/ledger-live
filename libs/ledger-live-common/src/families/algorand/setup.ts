// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-algorand/bridge/js";
import makeCliTools from "@ledgerhq/coin-algorand/cli-transaction";
import algorandResolver from "@ledgerhq/coin-algorand/hw-getAddress";
import { Transaction } from "@ledgerhq/coin-algorand/types";
import Algorand from "@ledgerhq/hw-app-algorand";
import Transport from "@ledgerhq/hw-transport";
import network from "@ledgerhq/live-network/network";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Algorand> = (transport: Transport) => {
  return new Algorand(transport);
};

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner), network);

const resolver: Resolver = createResolver(createSigner, algorandResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
