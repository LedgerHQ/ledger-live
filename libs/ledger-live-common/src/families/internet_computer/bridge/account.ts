import { makeAccountBridgeReceive, makeSync } from "../../../bridge/jsHelpers";
import { defaultUpdateTransaction } from "@ledgerhq/coin-framework/bridge/jsHelpers";
import {
  Account,
  AccountBridge,
  AccountLike,
  BroadcastFnSignature,
  SignOperationFnSignature,
} from "@ledgerhq/types-live";
import { InternetComputerOperation, Transaction, TransactionStatus } from "../types";
import { getAccountShape } from "./bridgeHelpers/account";
import BigNumber from "bignumber.js";
import { getEstimatedFees } from "./bridgeHelpers/fee";
import { getAddress, validateAddress, validateMemo } from "./bridgeHelpers/addresses";
import {
  AmountRequired,
  InvalidAddress,
  InvalidAddressBecauseDestinationIsAlsoSource,
  NotEnoughBalance,
  RecipientRequired,
} from "@ledgerhq/errors";
import { Observable } from "rxjs";
import { withDevice } from "../../../hw/deviceAccess";
import { encodeOperationId } from "../../../operation";
import {
  broadcastTxn,
  getTxnExpirationDate,
  getTxnMetadata,
  getUnsignedTransaction,
  signICPTransaction,
} from "./bridgeHelpers/icpRosetta";
import { getPath } from "../utils";
import { InvalidMemoICP } from "../errors";

const sync = makeSync({ getAccountShape });

const createTransaction = (): Transaction => {
  // log("debug", "[createTransaction] creating base tx");
  return {
    family: "internet_computer",
    amount: new BigNumber(0),
    fees: getEstimatedFees(),
    recipient: "",
    useAllAmount: false,
  };
};

const prepareTransaction = async (a: Account, t: Transaction): Promise<Transaction> => {
  // log("debug", "[prepareTransaction] start fn");

  const { address } = getAddress(a);
  const { recipient } = t;

  let amount: BigNumber = t.amount;
  if (recipient && address) {
    // log("debug", "[prepareTransaction] fetching estimated fees");

    if ((await validateAddress(recipient)).isValid && (await validateAddress(address)).isValid) {
      if (t.useAllAmount) {
        amount = a.spendableBalance.minus(t.fees);
        return { ...t, amount };
      }
    }
  }

  // log("debug", "[prepareTransaction] finish fn");
  return t;
};

const getTransactionStatus = async (a: Account, t: Transaction): Promise<TransactionStatus> => {
  const errors: TransactionStatus["errors"] = {};
  const warnings: TransactionStatus["warnings"] = {};

  const { balance } = a;
  const { address } = getAddress(a);
  const { recipient, useAllAmount } = t;
  let { amount } = t;

  if (!recipient) {
    errors.recipient = new RecipientRequired();
  } else if (!(await validateAddress(recipient)).isValid) {
    errors.recipient = new InvalidAddress("", {
      currencyName: a.currency.name,
    });
  } else if (recipient.toLowerCase() === address.toLowerCase()) {
    errors.recipient = new InvalidAddressBecauseDestinationIsAlsoSource();
  }

  if (!(await validateAddress(address)).isValid) {
    errors.sender = new InvalidAddress("", {
      currencyName: a.currency.name,
    });
  }

  if (!validateMemo(t.memo).isValid) {
    errors.transaction = new InvalidMemoICP();
  }

  // This is the worst case scenario (the tx won't cost more than this value)
  const estimatedFees = t.fees;

  let totalSpent: BigNumber;

  if (useAllAmount) {
    totalSpent = a.spendableBalance;
    amount = totalSpent.minus(estimatedFees);
    if (amount.lte(0) || totalSpent.gt(balance)) {
      errors.amount = new NotEnoughBalance();
    }
  } else {
    totalSpent = amount.plus(estimatedFees);
    if (amount.eq(0)) {
      errors.amount = new AmountRequired();
    } else if (totalSpent.gt(a.spendableBalance)) {
      errors.amount = new NotEnoughBalance();
    }
  }

  // log("debug", "[getTransactionStatus] finish fn");

  return {
    errors,
    warnings,
    estimatedFees,
    amount,
    totalSpent,
  };
};

const estimateMaxSpendable = async ({
  account,
  transaction,
}: {
  account: AccountLike;
  parentAccount?: Account | null | undefined;
  transaction?: Transaction | null | undefined;
}): Promise<BigNumber> => {
  const balance = account.balance;

  let maxSpendable = balance.minus(transaction?.fees ?? getEstimatedFees());

  if (maxSpendable.isLessThan(0)) {
    maxSpendable = new BigNumber(0);
  }

  return maxSpendable;
};

const signOperation: SignOperationFnSignature<Transaction> = ({ account, deviceId, transaction }) =>
  withDevice(deviceId)(
    transport =>
      new Observable(o => {
        async function main() {
          // log("debug", "[signOperation] start fn");

          const { recipient, amount } = transaction;
          const { id: accountId, xpub } = account;
          const { address, derivationPath } = getAddress(account);
          const fee = transaction.fees;

          const { unsignedTxn, payloads } = await getUnsignedTransaction(transaction, account);

          o.next({
            type: "device-signature-requested",
          });

          const { signedTxn } = await signICPTransaction({
            unsignedTxn,
            transport,
            path: getPath(derivationPath),
            payloads,
            pubkey: xpub ?? "",
          });

          const { hash } = await getTxnMetadata(signedTxn);

          o.next({
            type: "device-signature-granted",
          });

          const operation: InternetComputerOperation = {
            id: encodeOperationId(accountId, hash, "OUT"),
            hash,
            type: "OUT",
            senders: [address],
            recipients: [recipient],
            accountId,
            value: amount.plus(fee),
            fee,
            blockHash: null,
            blockHeight: null,
            date: new Date(),
            extra: {
              memo: transaction.memo,
            },
          };

          o.next({
            type: "signed",
            signedOperation: {
              operation,
              signature: signedTxn,
              expirationDate: getTxnExpirationDate(unsignedTxn),
            },
          });
        }

        main().then(
          () => o.complete(),
          e => o.error(e),
        );
      }),
  );

const receive = makeAccountBridgeReceive();

const broadcast: BroadcastFnSignature = async ({ signedOperation: { signature, operation } }) => {
  // log("debug", "[broadcast] start fn");

  await broadcastTxn(signature);

  const result = { ...operation };

  return result;
};

export const accountBridge: AccountBridge<Transaction> = {
  sync,
  updateTransaction: defaultUpdateTransaction,
  createTransaction,
  prepareTransaction,
  getTransactionStatus,
  estimateMaxSpendable,
  receive,
  signOperation,
  broadcast,
};
