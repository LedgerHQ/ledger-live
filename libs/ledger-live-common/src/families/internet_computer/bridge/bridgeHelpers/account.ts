import { GetAccountShape } from "../../../../bridge/jsHelpers";
import { decodeAccountId, encodeAccountId } from "../../../../account";
import { log } from "@ledgerhq/logs";
import { fetchBalances, fetchBlockHeight, fetchTxns } from "./api";
import { flatMap } from "lodash";
import { Account } from "@ledgerhq/types-live";
import BigNumber from "bignumber.js";
import { ICPRosettaGetTxnsHistoryResponse } from "./icpRosetta/types";
import { ICP_FEES } from "../../consts";
import { encodeOperationId } from "../../../../operation";
import { normalizeEpochTimestamp } from "../../utils";
import { InternetComputerOperation } from "../../types";

const mapTxToOps = (accountId: string, address: string, fee = ICP_FEES) => {
  return (
    txInfo: ICPRosettaGetTxnsHistoryResponse["transactions"][0],
  ): InternetComputerOperation[] => {
    const ops: InternetComputerOperation[] = [];
    const ownerOperation = txInfo.transaction.operations.find(
      cur => cur.account.address === address,
    );
    const counterOperation = txInfo.transaction.operations.find(
      cur => cur.account.address !== address,
    );

    if (!ownerOperation || !counterOperation) return ops;

    const timeStamp = txInfo.transaction.metadata.timestamp;
    const amount = BigNumber(ownerOperation.amount.value);
    const hash = txInfo.transaction.transaction_identifier.hash;
    const fromAccount = amount.isPositive()
      ? counterOperation.account.address
      : ownerOperation.account.address;
    const toAccount = amount.isNegative()
      ? counterOperation.account.address
      : ownerOperation.account.address;
    const memo = txInfo.transaction.metadata.memo.toString();
    const blockHeight = txInfo.transaction.metadata.block_height;

    const date = new Date(normalizeEpochTimestamp(timeStamp.toString()));
    const value = amount.abs();
    const feeToUse = BigNumber(fee);

    const isSending = amount.isNegative();
    const isReceiving = amount.isPositive();

    if (isSending) {
      ops.push({
        id: encodeOperationId(accountId, hash, "OUT"),
        hash,
        type: "OUT",
        value: value.plus(feeToUse),
        fee: feeToUse,
        blockHeight,
        blockHash: null,
        accountId,
        senders: [fromAccount],
        recipients: [toAccount],
        date,
        extra: {
          memo,
        },
      });
    }

    if (isReceiving) {
      ops.push({
        id: encodeOperationId(accountId, hash, "IN"),
        hash,
        type: "IN",
        value,
        fee: feeToUse,
        blockHeight,
        blockHash: null,
        accountId,
        senders: [fromAccount],
        recipients: [toAccount],
        date,
        extra: {
          memo,
        },
      });
    }

    return ops;
  };
};

export const getAccountShape: GetAccountShape = async info => {
  const { address, currency, derivationMode, rest = {}, initialAccount } = info;
  const publicKey = reconciliatePublicKey(rest.publicKey, initialAccount);

  const accountId = encodeAccountId({
    type: "js",
    version: "2",
    currencyId: currency.id,
    xpubOrAddress: publicKey,
    derivationMode,
  });

  log("debug", `Generation account shape for ${address}`);

  const blockHeight = await fetchBlockHeight();
  const balanceResp = await fetchBalances(address);
  const balance = balanceResp.balances[0];

  const txns = await fetchTxns(address);
  const result: Partial<Account> = {
    id: accountId,
    balance: BigNumber(balance.value),
    spendableBalance: BigNumber(balance.value),
    operations: flatMap(txns.transactions.reverse(), mapTxToOps(accountId, address)),
    blockHeight: blockHeight.current_block_identifier.index,
    operationsCount: txns.transactions.length,
    xpub: publicKey,
  };

  return result;
};

function reconciliatePublicKey(publicKey?: string, initialAccount?: Account): string {
  if (publicKey) return publicKey;
  if (initialAccount) {
    const { xpubOrAddress } = decodeAccountId(initialAccount.id);
    return xpubOrAddress;
  }
  throw new Error("publicKey wasn't properly restored");
}
