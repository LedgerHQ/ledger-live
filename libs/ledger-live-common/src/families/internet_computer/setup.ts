// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-internet_computer/bridge/index";
import Transport from "@ledgerhq/hw-transport";
import ICP from "@zondax/ledger-icp";
import icpResolver from "@ledgerhq/coin-internet_computer/signer/index";
import { signMessage } from "@ledgerhq/coin-internet_computer/hw-signMessage";
import type { Account, Bridge } from "@ledgerhq/types-live";
import makeCliTools from "@ledgerhq/coin-internet_computer/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";
import { TransactionStatus, Transaction } from "@ledgerhq/coin-internet_computer/types/index";
import { ICPGetAddrResponse, ICPSignature, ICPSigner } from "./types";
import { getPath, isError } from "./common";

const createSigner: CreateSigner<ICPSigner> = (transport: Transport) => {
  const icp = new ICP(transport);
  return {
    showAddressAndPubKey: async (path: string): Promise<ICPGetAddrResponse> => {
      const r = await icp.showAddressAndPubKey(getPath(path));
      isError(r);

      return r;
    },
    getAddressAndPubKey: async (path: string): Promise<ICPGetAddrResponse> => {
      const r = await icp.getAddressAndPubKey(getPath(path));
      isError(r);

      return r;
    },
    sign: async (path: string, message: Buffer): Promise<ICPSignature> => {
      const r = await icp.sign(getPath(path), message, 0);
      isError(r);

      return r;
    },
  };
};

const bridge: Bridge<Transaction, Account, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
);

const messageSigner = {
  signMessage,
};

const resolver: Resolver = createResolver(createSigner, icpResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, messageSigner, resolver };
