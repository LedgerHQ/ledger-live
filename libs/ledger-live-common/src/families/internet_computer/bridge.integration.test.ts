import "../../__tests__/test-helpers/setup";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import type { CurrenciesData, DatasetTest } from "@ledgerhq/types-live";
import type { Transaction } from "./types";
import { fromTransactionRaw } from "./transaction";
import BigNumber from "bignumber.js";
import { AmountRequired, InvalidAddress, NotEnoughBalance } from "@ledgerhq/errors";
import { getEstimatedFees } from "./bridge/bridgeHelpers/fee";
import { InvalidMemoICP } from "./errors";

// const SEED_IDENTIFIER =
// "e8a1474afbed438be8b019c4293b9e01b33075d72757ac715183ae7c7ba77e37";
const ACCOUNT_2 = "fdb7db0d3ae67368cb5010b7de7d98566c072f0a4eda871f45cd6582bf08aeb4";

const internet_computer: CurrenciesData<Transaction> = {
  scanAccounts: [
    {
      name: "internet_computer seed 1",
      apdus: `
      => 11010000142c000080df000080000000800000000000000000
      <= 046f08828871028b6e3cb5c13b2e2a8fa6e93f0b3ca7379171f6b7b45877955a2430925f76ec69ccb3cd8738859a8e29dcd0f9a357f1d009d2b497c6c8f63aa7cf182d2af0e048c8cf215fa86d7af7e2234f401679c7839dd3b1ae916a02e8a1474afbed438be8b019c4293b9e01b33075d72757ac715183ae7c7ba77e37793471646770797966757670627963697a646873637835696e763570707972646a3561626d366f68716f6f35686d6e6f73667661659000
      => 11010000142c000080df000080000000800000000001000000
      <= 043ad9dda46b25cbcf98b2d91f8aa289d08078fa960d13e7d77a571c625eedcb62b6c26a86408d30a6bbdea5ecad6f7603bc1ba11fbb62caa315789f333ece2c8cd785035623de419beb00247d11fbb4d3f77fbb466869fb01deac4f61021723b9a25b7cb1865caa6e633fc727c95cee6a84c52e5ca0f8752ecf66cdea4b6764756f79346f78717562766d69363669676e367761626570756937786e677436353733777274696e6835716478766d6a357171659000
      `,
      test: (expect, accounts) => {
        for (const account of accounts) {
          expect(account.derivationMode).toEqual("internet_computer");
        }
      },
    },
  ],
  accounts: [
    {
      raw: {
        balance: "1000000",
        currencyId: "internet_computer",
        derivationMode: "internet_computer",
        freshAddress: "e8a1474afbed438be8b019c4293b9e01b33075d72757ac715183ae7c7ba77e37",
        freshAddressPath: "44'/223'/0'/0/0",
        freshAddresses: [
          {
            address: "e8a1474afbed438be8b019c4293b9e01b33075d72757ac715183ae7c7ba77e37",
            derivationPath: "44'/223'/0'/0/0",
          },
        ],
        id: "js:2:internet_computer:046f08828871028b6e3cb5c13b2e2a8fa6e93f0b3ca7379171f6b7b45877955a2430925f76ec69ccb3cd8738859a8e29dcd0f9a357f1d009d2b497c6c8f63aa7cf:",
        index: 0,
        name: "Internet Computer 1",
        nfts: undefined,
        operationsCount: 1,
        blockHeight: 0,
        pendingOperations: [],
        operations: [],
        lastSyncDate: "",
        seedIdentifier:
          "046f08828871028b6e3cb5c13b2e2a8fa6e93f0b3ca7379171f6b7b45877955a2430925f76ec69ccb3cd8738859a8e29dcd0f9a357f1d009d2b497c6c8f63aa7cf",
        spendableBalance: "1000000",
        starred: false,
        swapHistory: [],
        syncHash: undefined,
        unitMagnitude: 8,
        used: true,
        xpub: "046f08828871028b6e3cb5c13b2e2a8fa6e93f0b3ca7379171f6b7b45877955a2430925f76ec69ccb3cd8738859a8e29dcd0f9a357f1d009d2b497c6c8f63aa7cf",
      },
      transactions: [
        {
          name: "Not a valid address",
          transaction: fromTransactionRaw({
            family: "internet_computer",
            recipient: "novalidaddress",
            fees: getEstimatedFees().toString(),
            amount: "1000",
          }),
          expectedStatus: {
            errors: {
              recipient: new InvalidAddress(),
            },
            warnings: {},
          },
        },
        {
          name: "Not enough balance",
          transaction: fromTransactionRaw({
            family: "internet_computer",
            recipient: ACCOUNT_2,
            fees: getEstimatedFees().toString(),
            amount: (300 * 1e9).toString(),
          }),
          expectedStatus: {
            errors: {
              amount: new NotEnoughBalance(),
            },
            warnings: {},
          },
        },
        {
          name: "Invalid transferID/Memo",
          transaction: fromTransactionRaw({
            family: "internet_computer",
            recipient: ACCOUNT_2,
            fees: getEstimatedFees().toString(),
            amount: "1000",
            memo: "-1",
          }),
          expectedStatus: {
            errors: {
              transaction: new InvalidMemoICP(),
            },
            warnings: {},
          },
        },
        {
          name: "Amount Required",
          transaction: fromTransactionRaw({
            family: "internet_computer",
            recipient: ACCOUNT_2,
            amount: "0",
            fees: getEstimatedFees().toString(),
          }),
          expectedStatus: {
            errors: {
              amount: new AmountRequired(),
            },
            warnings: {},
          },
        },
        {
          name: "New account and sufficient amount",
          transaction: fromTransactionRaw({
            family: "internet_computer",
            recipient: ACCOUNT_2,
            amount: "1000",
            fees: getEstimatedFees().toString(),
          }),
          expectedStatus: {
            amount: new BigNumber("1000"),
            errors: {},
            warnings: {},
          },
        },
      ],
    },
  ],
};

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    internet_computer,
  },
};

testBridge(dataset);
