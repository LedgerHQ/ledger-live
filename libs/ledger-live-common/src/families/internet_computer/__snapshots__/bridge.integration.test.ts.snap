// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`internet_computer currency bridge scanAccounts internet_computer seed 1 1`] = `
Array [
  Object {
    "balance": "1000000",
    "currencyId": "internet_computer",
    "derivationMode": "internet_computer",
    "freshAddress": "e8a1474afbed438be8b019c4293b9e01b33075d72757ac715183ae7c7ba77e37",
    "freshAddressPath": "44'/223'/0'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "e8a1474afbed438be8b019c4293b9e01b33075d72757ac715183ae7c7ba77e37",
        "derivationPath": "44'/223'/0'/0/0",
      },
    ],
    "id": "js:2:internet_computer:046f08828871028b6e3cb5c13b2e2a8fa6e93f0b3ca7379171f6b7b45877955a2430925f76ec69ccb3cd8738859a8e29dcd0f9a357f1d009d2b497c6c8f63aa7cf:internet_computer",
    "index": 0,
    "name": "Internet Computer 1",
    "nfts": undefined,
    "operationsCount": 1,
    "pendingOperations": Array [],
    "seedIdentifier": "046f08828871028b6e3cb5c13b2e2a8fa6e93f0b3ca7379171f6b7b45877955a2430925f76ec69ccb3cd8738859a8e29dcd0f9a357f1d009d2b497c6c8f63aa7cf",
    "spendableBalance": "1000000",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 8,
    "used": true,
    "xpub": "046f08828871028b6e3cb5c13b2e2a8fa6e93f0b3ca7379171f6b7b45877955a2430925f76ec69ccb3cd8738859a8e29dcd0f9a357f1d009d2b497c6c8f63aa7cf",
  },
  Object {
    "balance": "0",
    "currencyId": "internet_computer",
    "derivationMode": "internet_computer",
    "freshAddress": "1723b9a25b7cb1865caa6e633fc727c95cee6a84c52e5ca0f8752ecf66cdea4b",
    "freshAddressPath": "44'/223'/0'/0/1",
    "freshAddresses": Array [
      Object {
        "address": "1723b9a25b7cb1865caa6e633fc727c95cee6a84c52e5ca0f8752ecf66cdea4b",
        "derivationPath": "44'/223'/0'/0/1",
      },
    ],
    "id": "js:2:internet_computer:043ad9dda46b25cbcf98b2d91f8aa289d08078fa960d13e7d77a571c625eedcb62b6c26a86408d30a6bbdea5ecad6f7603bc1ba11fbb62caa315789f333ece2c8c:internet_computer",
    "index": 1,
    "name": "Internet Computer 2",
    "nfts": undefined,
    "operationsCount": 0,
    "pendingOperations": Array [],
    "seedIdentifier": "046f08828871028b6e3cb5c13b2e2a8fa6e93f0b3ca7379171f6b7b45877955a2430925f76ec69ccb3cd8738859a8e29dcd0f9a357f1d009d2b497c6c8f63aa7cf",
    "spendableBalance": "0",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 8,
    "used": false,
    "xpub": "043ad9dda46b25cbcf98b2d91f8aa289d08078fa960d13e7d77a571c625eedcb62b6c26a86408d30a6bbdea5ecad6f7603bc1ba11fbb62caa315789f333ece2c8c",
  },
]
`;

exports[`internet_computer currency bridge scanAccounts internet_computer seed 1 2`] = `
Array [
  Array [
    Object {
      "accountId": "js:2:internet_computer:046f08828871028b6e3cb5c13b2e2a8fa6e93f0b3ca7379171f6b7b45877955a2430925f76ec69ccb3cd8738859a8e29dcd0f9a357f1d009d2b497c6c8f63aa7cf:internet_computer",
      "blockHash": null,
      "blockHeight": 5432782,
      "contract": undefined,
      "extra": Object {
        "memo": "0",
      },
      "fee": "10000",
      "hash": "c63c65091a4de0b9aec1b8bd242fa97b96dc8a7349577470f788c1f3df6fd7b8",
      "id": "js:2:internet_computer:046f08828871028b6e3cb5c13b2e2a8fa6e93f0b3ca7379171f6b7b45877955a2430925f76ec69ccb3cd8738859a8e29dcd0f9a357f1d009d2b497c6c8f63aa7cf:internet_computer-c63c65091a4de0b9aec1b8bd242fa97b96dc8a7349577470f788c1f3df6fd7b8-IN",
      "operator": undefined,
      "recipients": Array [
        "e8a1474afbed438be8b019c4293b9e01b33075d72757ac715183ae7c7ba77e37",
      ],
      "senders": Array [
        "fdb7db0d3ae67368cb5010b7de7d98566c072f0a4eda871f45cd6582bf08aeb4",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "1000000",
    },
  ],
  Array [],
]
`;
